### PR TITLE
MAP-258 Allow cell move for new cell allocation api

### PR DESCRIPTION
**Pull request by Mjwillis**
- Remove service annotations checks an move to resource classes
- Added new MAINTAIN_CELL_MOVEMENTS role for API use

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingMovementsResource.java [hunk 1]
```diff
@@ -120,6 +122,14 @@ public CellMoveResult moveToCell(@PathVariable("bookingId") @Parameter(descripti
             @ApiResponse(responseCode = "404", description = "Requested resource not found.", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))}),
             @ApiResponse(responseCode = "500", description = "Unrecoverable error occurred whilst processing request.", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))})})
     @PutMapping("/{bookingId}/move-to-cell-swap")
+    @Operation(
+        summary = "Move the prisoner from current cell to cell swap",
+        description = "Using role MAINTAIN_CELL_MOVEMENTS will no longer check for user access to prisoner booking, this endpoint will be removed in future releases"
+    )
+    @ProxyUser
+    @Deprecated
+    @VerifyBookingAccess(overrideRoles = {"MAINTAIN_CELL_MOVEMENTS"})
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingMovementsResource.java:131]: Moved here from service layer

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderMovementsResource.kt [hunk 2]
```diff
@@ -44,6 +46,8 @@ class OffenderMovementsResource(
   @Tag(name = "unilink")
   @PutMapping("/{offenderNo}/living-unit/{internalLocationDescription}")
   @ProxyUser
+  @VerifyOffenderAccess(overrideRoles = ["SYSTEM_USER", "UNILINK", "MAINTAIN_CELL_MOVEMENTS"])
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderMovementsResource.kt:49]: Moved to resource layer and used offender no verification instead

File: src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java [hunk 3]
```diff
@@ -46,8 +46,6 @@ public BedAssignmentHistoryService(
         this.maxBatchSize = maxBatchSize;
     }
 
-    @VerifyBookingAccess(overrideRoles = {"SYSTEM_USER", "UNILINK"})
-    @HasWriteScope
     public BedAssignmentHistoryPK add(final Long bookingId, final Long livingUnitId, final String reasonCode, final LocalDateTime time) {
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java:49]: This is only called from one service via one endpoint which has the protection

File: src/main/java/uk/gov/justice/hmpps/prison/service/BookingService.java [hunk 4]
```diff
@@ -781,8 +781,6 @@ public void updateLivingUnit(final Long bookingId, final String livingUnitDescri
     }
 
     @Transactional
-    @VerifyBookingAccess(overrideRoles = {"SYSTEM_USER", "UNILINK"})
-    @HasWriteScope
     public void updateLivingUnit(final Long bookingId, final AgencyInternalLocation location) {
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/BookingService.java:784]: This is only called from one service via one endpoint which has the protection

File: src/main/java/uk/gov/justice/hmpps/prison/service/MovementUpdateService.java [hunk 5]
```diff
@@ -45,8 +45,6 @@ public MovementUpdateService(
     }
 
     @Transactional
-    @VerifyBookingAccess(overrideRoles = {"SYSTEM_USER", "UNILINK"})
-    @HasWriteScope
     public CellMoveResult moveToCellOrReception(final Long bookingId, final String internalLocationDescription, final String reasonCode, final LocalDateTime dateTime) {
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/MovementUpdateService.java:48]: This is only called from one service via one endpoint which has the protection

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingMovementsResource.java [hunk 6]
```diff
@@ -104,6 +106,8 @@ public CourtHearings getCourtHearings(@PathVariable("bookingId") @Parameter(desc
             @ApiResponse(responseCode = "500", description = "Unrecoverable error occurred whilst processing request.", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))})})
     @PutMapping("/{bookingId}/living-unit/{internalLocationDescription}")
     @ProxyUser
+    @VerifyBookingAccess(overrideRoles = {"SYSTEM_USER", "MAINTAIN_CELL_MOVEMENTS"})
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingMovementsResource.java:109]: Moved up

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingMovementsResourceIntTest_moveToCellSwap.java [hunk 7]
```diff
@@ -164,6 +175,17 @@ private String differentAgencyToken() {
         );
     }
 
+    private String noUserInContext() {
+        return jwtAuthenticationHelper.createJwt(
+            JwtParameters.builder()
+                .username("a-system-client-id")
+                .scope(List.of("read", "write"))
+                .roles(List.of("MAINTAIN_CELL_MOVEMENTS"))
```
Comment (Mjwillis) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingMovementsResourceIntTest_moveToCellSwap.java:183]: This has the role needed to by-pass

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**